opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java

```diff
@@ -84,43 +83,88 @@ public static Tracer get() {
         return INSTANCE;
     }
 
+    /**
+     * Identify whether a {@link Tracer} has previously been registered.
+     * <p>
+     * This check is useful in scenarios where more than one component may be responsible
+     * for registering a tracer. For example, when using a Java Agent, it will need to determine
+     * if the application has already registered a tracer, and if not attempt to resolve and
+     * register one itself.
+     *
+     * @return Whether a tracer has been registered
+     */
+    public static synchronized boolean isRegistered() {
+        return !(GlobalTracer.tracer instanceof NoopTracer);
+    }
+
     /**
      * Register a {@link Tracer} to back the behaviour of the {@link #get() global tracer}.
      * <p>
-     * Registration is a one-time operation, attempting to call it more often will result in a runtime exception.
+     * The tracer is provided through a supplier that will only be called if global tracer is currently absent.
+     * Registration is a one-time operation. Once a tracer has been registered, all attempts at re-registering
+     * will return {@code false}.
      * <p>
      * Every application intending to use the global tracer is responsible for registering it once
      * during its initialization.
      *
-     * @param tracer Tracer to use as global tracer.
-     * @throws RuntimeException if there is already a current tracer registered
+     * @param supplier Supplier for the tracer to use as global tracer.
+     * @return {@code true} if the supplied tracer was registered as a result of this call,
+     * {@code false} otherwise.
      */
-    public static synchronized void register(final Tracer tracer) {
-        if (tracer == null) {
-            throw new NullPointerException("Cannot register GlobalTracer <null>.");
-        }
-        if (tracer instanceof GlobalTracer) {
-            LOGGER.log(Level.FINE, "Attempted to register the GlobalTracer as delegate of itself.");
-            return; // no-op
+    public static synchronized boolean registerIfAbsent(final TracerSupplier supplier) {
+        if (supplier != null && !isRegistered()) {
+            final Tracer suppliedTracer = supplier.get();
+            if (suppliedTracer != null && !(suppliedTracer instanceof GlobalTracer)) {
+                GlobalTracer.tracer = suppliedTracer;
+                return true;
```

Comment:
what is the point of null checks? This being a configuration step I'd rather see NPE than silent success.

Reply:
The null checks are there because the new API returns `true` for registration and `false` for _no registration_, which seems a valid way to handle this.
Just my opinion though, there is something to be said for throwing NPE's as well.

Reply:
true/false for registration or no registration is fine, but null supplier or tracer are violations of the method's API, imo, it's better to report them than work around them.

Reply:
Valid point. I'd want a NPE _with_ a message what was `null` though. Unfortunately we don't have `Objects.requireNonNull` in java 6 yet..

Reply:
`if (supplier == null) throw new NullPointerException("supplier must not be null");` no bueno?